### PR TITLE
Rootless bridge: preserve source IPs via pesto/pasta

### LINKED ISSUE #8193: Alternate port_handler that keeps the source ip for user-defined rootless networks
**Is this a BUG REPORT or FEATURE REQUEST? (leave only one on its own line)**

/kind bug

**Description**

Alternate "port_handler=slirt4netns" provided in PR 6965 (https://github.com/containers/podman/pull/6965) for "only 127.0.0.1 within containers" **not implemented** for user-defined rootless cni networks
- specifying "port_handler=slirp4netns" when **connecting** a container to a already-created user-defined rootless cni network does **NOT** error out and does **NOT** implement the behavior provided in PR 6965
- as a result, not possible to see accurate remote address within containers connected to a user-defined rootless cni network (only "127.0.0.1")

**Steps to reproduce the issue:**

```
normaluser@containerhost $> podman network create myCNI

normaluser@containerhost $> podman run --name myNginx1 --publish 8081:80 --network=myCNI -d nginx:alpine
normaluser@containerhost $> podman run --name myNginx2 --publish 8082:80 --network=myCNI:port_handler=slirp4netns -d nginx:alpine
normaluser@containerhost $> podman run --name myNginx3 --publish 8083:80 --network=slirp4netns:port_handler=slirp4netns -d nginx:alpine

normaluser@containerhost $>  podman inspect myNginx1 | grep -i ipaddress
                "IPAddress": "10.89.0.2",
normaluser@containerhost $>  podman inspect myNginx2 | grep -i ipaddress
                "IPAddress": "",
normaluser@containerhost $>  podman inspect myNginx3 | grep -i ipaddress
                "IPAddress": "",

normaluser@containerhost $> podman exec myNginx1 ifconfig
                ## shows loopback and **eth0**, as expected
normaluser@containerhost $> podman exec myNginx2 ifconfig
                ## absolutely blank, not expected
normaluser@containerhost $> podman exec myNginx3 ifconfig
                ## shows loopback and **tap0**, as expected

otherhost $> curl --head http://_containerhost_:8081
otherhost $> curl --head http://_containerhost_:8082
`        curl: (7) Failed to connect to _containerhost_ port 8082: Connection refused`
otherhost $> curl --head http://_containerhost_:8083

normaluser@containerhost $> podman logs myNginx1 | grep HEAD
        127.0.0.1 - - [29/Oct/2020:15:51:25 -0700] "HEAD / HTTP/1.1" 200 0 "-" "curl/7.69.1" "-"
normaluser@containerhost $> podman logs myNginx2 | grep HEAD
        ## nothing to see, no connection succeeded
normaluser@containerhost $> podman logs myNginx3|grep HEAD
        192.168.0.116 - - [29/Oct/2020:15:51:31 -0700] "HEAD / HTTP/1.1" 200 0 "-" "curl/7.69.1" "-"
```
**Describe the results you received:**
"myNginx1" container has only 127.0.0.1 as remote address (confusing), though explained in PR 6965
"myNginx2" container launches without error **however** port not opened and no ip address assigned
"myNginx3" container - no problem - everything makes sense

**Describe the results you expected:**
"myNginx1:" expected to see correct remote address wihin container (though explainded in PR 6965)
"myNginx2:" expected successful ip assignment and port to open using "port_handler=slirp4netns", or contaier launch to fail
"myNginx3" container - no problem - everything makes sense

**Additional information you deem important (e.g. issue happens only occasionally):**
happens consistently and reproducably

**Output of `podman version`:**

```
        Version:      2.1.1
        API Version:  2.0.0
        Go Version:   go1.14
        Built:        Wed Dec 31 16:00:00 1969
        OS/Arch:      linux/amd64
```

**Output of `podman info --debug`:**

```
host:
  arch: amd64
  buildahVersion: 1.16.1
  cgroupManager: cgroupfs
  cgroupVersion: v1
  conmon:
    package: 'conmon: /usr/libexec/podman/conmon'
    path: /usr/libexec/podman/conmon
    version: 'conmon version 2.0.20, commit: '
  cpus: 4
  distribution:
    distribution: debian
    version: "10"
  eventLogger: journald
  hostname: arachne
  idMappings:
    gidmap:
    - container_id: 0
      host_id: 2000
      size: 1
    - container_id: 1
      host_id: 100001
      size: 65536
    uidmap:
    - container_id: 0
      host_id: 2000
      size: 1
    - container_id: 1
      host_id: 100001
      size: 65536
  kernel: 4.19.0-10-amd64
  linkmode: dynamic
  memFree: 369057792
  memTotal: 2091732992
  ociRuntime:
    name: runc
    package: 'runc: /usr/sbin/runc'
    path: /usr/sbin/runc
    version: |-
      runc version 1.0.0~rc6+dfsg1
      commit: 1.0.0~rc6+dfsg1-3
      spec: 1.0.1
  os: linux
  remoteSocket:
    path: /run/user/2000/podman/podman.sock
  rootless: true
  slirp4netns:
    executable: /usr/bin/slirp4netns
    package: 'slirp4netns: /usr/bin/slirp4netns'
    version: |-
      slirp4netns version 1.1.4
      commit: unknown
      libslirp: 4.3.1-git
      SLIRP_CONFIG_VERSION_MAX: 3
  swapFree: 746057728
  swapTotal: 1073737728
  uptime: 184h 14m 26.29s (Approximately 7.67 days)
registries:
  search:
  - docker.io
  - quay.io
version:
  APIVersion: 2.0.0
  Built: 0
  BuiltTime: Wed Dec 31 16:00:00 1969
  GitCommit: ""
  GoVersion: go1.14
  OsArch: linux/amd64
  Version: 2.1.1
```

**Package info (e.g. output of `rpm -q podman` or `apt list podman`):**

```
podman/unknown,now 2.1.1~2 amd64 [installed]
podman/unknown 2.1.1~2 arm64
podman/unknown 2.1.1~2 armhf
podman/unknown 2.1.1~2 ppc64el
```

**Have you tested with the latest version of Podman and have you checked the Podman Troubleshooting Guide?**

Yes

**Additional environment details (AWS, VirtualBox, physical, etc.):**
Container host is a VirtualBox VM running on Fedora 32
podman packages installed from OpenSuse repo
**$> cat "/etc/apt/sources.list.d/devel:kubic:libcontainers:stable.list"**
```
## "Kubic" repo from "OpenSuse" for "podman" packages since they aren't in Debian 10 (buster) repos
deb https://download.opensuse.org/repositories/devel:/kubic:/libcontainers:/stable/Debian_10/ /
```
**$> cat /etc/os-release** 
`PRETTY_NAME="Debian GNU/Linux 10 (buster)"`

### PR DESCRIPTION
**Pull request by Honny1**
# Require passta version: 

This requires the `pesto` binary, available since `passt-0^20260507.g1afd4ed`. 

## Local Passt Setup + testing

```bash
# Build pasta with pesto support
git clone git://passt.top/passt 
cd passt
git checkout ec96f0124282 
b4 shazam https://archives.passt.top/passt-dev/20260506092241.1607480-1-sbrivio@redhat.com/
make 
sudo make install

cd ../podman
make binaries
make localintegration FOCUS_FILE=run_networking_test.go
```

# TODO:
- [ ] Proper vendor container-libs
- [x] Rebase + drop commit after merging removal of slirp (https://github.com/containers/podman/pull/27828)
- [x] Documentation
- [x] Deal with the strict pasta version question.  (https://github.com/containers/container-libs/pull/860)

----

Add `rootless_port_forwarder="pasta"` option in `containers.conf` that switches rootless bridge port forwarding from `rootlessport` to `pasta`, preserving source IPs. The default remains `rootlessport`.

### Problem

When running rootless containers on a bridge network (`podman run -p 8080:80 --network mynet`), the old `rootlessport` userspace TCP/UDP proxy destroyed source IP information. Every connection appeared to come from `127.0.0.1` inside the container, regardless of the actual client IP.

### Solution

Pesto is a client tool for `passta` that dynamically updates port forwarding rules via a UNIX domain socket. Instead of proxying traffic in userspace (which loses source IPs), pesto configures `pasta` to forward at the kernel level using splice (localhost) or TAP (external traffic), preserving the original source IP.

### How it works

```
Host -> pasta (splice/TAP) -> rootless netns -> netavark bridge DNAT -> container
```

1. A shared `pasta` instance runs in the rootless network namespace with a control socket (`-c pasta.sock`)
2. On **container start**: netavark sets up the bridge + nftables DNAT rules, then `pesto --add` registers ports in pasta's forwarding table
3. On **container stop**: `pesto --delete` removes the container's ports, then netavark tears down bridge/DNAT
4. On **network connect/disconnect/reload**: port forwarding state is updated accordingly

Key implementation details:

- **HostIP stripping in the netavark wrapper** (`convertNetOpts`): when `rootless_port_forwarder="pasta"` is active, HostIP is stripped from port mappings before passing to netavark. Pesto handles host-side address binding; netavark's DNAT rules inside the rootless netns must not restrict on destination address since pasta's splice delivers traffic with a different destination than the user-specified HostIP
- **Dual-stack binding**: when no explicit HostIP is set, pesto binds both `0.0.0.0` and `[::]` so IPv6 networks work out of the box
- **Config toggle**: `rootless_port_forwarder` in `[network]` section of `containers.conf` selects between `rootlessport` (default) and `pasta`

### Current limitations

- Same port on different HostIPs (e.g. `-p 127.0.0.1:8080:80` and `-p 127.0.0.2:8080:80` on separate containers) does not work in pasta mode because HostIP stripping causes conflicting DNAT rules. This requires destination mapping support in pesto.

Fixes: https://redhat.atlassian.net/browse/RUN-2214
Fixes: https://github.com/containers/podman/issues/8193
Fixes: https://redhat.atlassian.net/browse/RUN-3587

Depends on: `passt >= passt-0^20260507.g1afd4ed`
Depends on: https://github.com/containers/container-libs/pull/755
Depends on: https://github.com/containers/podman/pull/27828
Depends on: https://github.com/containers/podman/pull/28451

#### Checklist

Ensure you have completed the following checklist for your pull request to be reviewed:

- [x] Certify you wrote the patch or otherwise have the right to pass it on as an open-source patch by signing all
commits. (`git commit -s`). (If needed, use `git commit -s --amend`).  The author email must match
the sign-off email address. See [CONTRIBUTING.md](https://github.com/containers/podman/blob/main/CONTRIBUTING.md#sign-your-prs)
for more information.
- [x] Referenced issues using `Fixes: #00000` in commit message (if applicable)
- [x] [Tests](https://github.com/containers/podman/tree/main/test#readme) have been added/updated (or no tests are needed)
- [x] [Documentation](https://github.com/containers/podman/blob/main/docs/README.md) has been updated (or no documentation changes are needed)
- [x] All commits pass `make validatepr` (format/lint checks)
- [x] [Release note](https://github.com/kubernetes/community/blob/master/contributors/guide/release-notes.md) entered in the section below (or `None` if no user-facing changes)

#### Does this PR introduce a user-facing change?

```release-note
Add experimental rootless_port_forwarder="pasta" option in containers.conf. When enabled, rootless bridge network port forwarding uses pasta's kernel-level forwarding (via pesto) instead of rootlessport's userspace proxy, preserving the original client source IP inside containers. Requires a passt build with pesto support. The default remains rootlessport.
```

### CODE REVIEW

**Comment by Honny1** [comment]
Question for reviewers: Should the option for rootlessport exist, or can it be dropped fully?

File: docs/source/markdown/options/network.md [hunk 1]
```diff
@@ -35,6 +35,8 @@ Valid _mode_ values are:
 
     Any other options will be passed through to netavark without validation. This can be useful to pass arguments to netavark plugins.
 
+    For rootless bridge networks, port forwarding uses `rootlessport` by default. Setting `rootless_port_forwarder="pasta"` in the `[network]` section of **[containers.conf(5)](https://github.com/containers/container-libs/blob/main/common/docs/containers.conf.5.md)** switches to pasta's kernel-level forwarding (via `pesto`), which preserves the original client source IP address inside the container.
```
Comment (Luap99) [inline docs/source/markdown/options/network.md]: which should ensure to document the exact pasta behavior is still experimental and subject to change

File: libpod/networking_linux.go [hunk 2]
```diff
@@ -59,15 +59,13 @@ func (r *Runtime) configureNetNS(ctr *Container, ctrNS string) (status map[strin
...
-		// make sure to fix this in container.handleRestartPolicy() as well
-		// Important we have to call this after r.setUpNetwork() so that
-		// we can use the proper netStatus
-		err = r.setupRootlessPortMappingViaRLK(ctr, ctrNS, netStatus)
+	// Set up port forwarding for rootless bridge networks.
+	if rootless.IsRootless() && len(ctr.config.PortMappings) > 0 {
+		if r.config.Network.RootlessPortForwarder == "pasta" {
+			err = r.setupRootlessPortMappingViaPesto(ctr)
+		} else {
+			err = r.setupRootlessPortMappingViaRLK(ctr, ctrNS, netStatus)
```
Comment (Luap99) [inline libpod/networking_linux.go]: I would have the switch case over r.config.Network.RootlessPortForwarder here and error on invalid value instead of the main containers.conf validation in common, see my other comment there

File: libpod/oci_conmon_common.go [hunk 3]
```diff
@@ -1188,8 +1188,11 @@ func (r *ConmonOCIRuntime) createOCIContainer(ctr *Container, restoreOptions *Co
 		// process cannot use them.
 		cmd.ExtraFiles = append(cmd.ExtraFiles, ports...)
 
-		// For rootless port forwarding, create sync pipe and leak write end to conmon
-		if rootless.IsRootless() && len(ctr.config.PortMappings) > 0 {
+		// For rootless port forwarding via rootlessport, create sync pipe and
+		// leak write end to conmon. Pasta forwarding mode does not use
+		// rootlessport, so no pipe is needed.
+		if rootless.IsRootless() && len(ctr.config.PortMappings) > 0 &&
+			ctr.runtime.config.Network.RootlessPortForwarder != "pasta" {
```
Comment (Luap99) [inline libpod/oci_conmon_common.go]: having these branched to be ` == "rootlessport"` would seem more logic to readers.
Also overall it would be helpful to define both strings as content and use them consistently then to avoid typos and allow code readers easily check of where they are referenced

File: test/e2e/common_test.go [hunk 4]
```diff
@@ -1629,6 +1629,56 @@ func testPortConnection(port int) {
 	Expect(err).ToNot(HaveOccurred())
 }
 
+// sendMessageToAddr sends a message to the given tcp address (host:port).
+//
+//nolint:unused,nolintlint // only called from linux-only test files, unused on freebsd
```
Comment (Luap99) [inline test/e2e/common_test.go]: well then put them into the file where they are used instead of this no lint thing

File: test/e2e/run_networking_test.go [hunk 5]
```diff
@@ -1049,6 +1051,456 @@ EXPOSE 2004-2005/tcp`, ALPINE)
...
+			// (source == bridge gateway) is gone.
+			//
+			// The exact mapped address is a pasta implementation detail.
+			Expect(output).ToNot(MatchRegexp(`connect to .* from .*` + subnetPrefix))
+		} else {
+			Expect(output).To(MatchRegexp(`connect to .* from .*` + subnetPrefix))
+		}
+	}
+
+	It("podman run bridge network preserves source IP", func() {
```
Comment (Luap99) [inline test/e2e/run_networking_test.go]: I am not sure about all of these tests

for source ip we have `podman networking: port with --userns=keep-id for rootless or --uidmap=* for rootful` system test

And in general we have a ton of other networking tests already, I rather not duplicate a bunch of slightly different cases.

If we want to test this then we likely want to create a loop over selected existing network tests that run twice with both rootlessport/pasta

File: test/e2e/run_networking_test.go [hunk 6]
```diff
@@ -1049,6 +1051,456 @@ EXPOSE 2004-2005/tcp`, ALPINE)
...
+			// rather than matching an exact IP because pasta maps localhost traffic
+			// to a host virtual IP (e.g. 192.168.64.2) whose value depends on the
+			// pasta version, host network config, and whether the traffic arrives
+			// via splice (localhost) or TAP (external).
+			//
+			// The important invariant is that the old rootlessport proxy behavior
+			// (source == bridge gateway) is gone.
+			//
+			// The exact mapped address is a pasta implementation detail.
+			Expect(output).ToNot(MatchRegexp(`connect to .* from .*` + subnetPrefix))
```
Comment (Luap99) [inline test/e2e/run_networking_test.go]: that does not seem to match this though..., it is the exact inverse from the other case below...

**Comment by Honny1** [comment]
PTAL @Luap99 @containers/podman-maintainers 

I am not sure about the packit failure. 

cc @sbrivio-rh

**Comment by sbrivio-rh** [comment]
> I am not sure about the packit failure.

To me the Packit failure (https://dashboard.packit.dev/jobs/testing-farm/1544417 / https://artifacts.dev.testing-farm.io/177feab7-7054-4ff9-9cf3-2f9f3f9ac985/) looks like a failure of a relevant test. From the second link:

```
not ok 619 [500] podman network reload
# (from function `bail-now' in file ./helpers.bash, line 230,
#  from function `die' in file ./helpers.bash, line 967,
#  from function `run_podman' in file ./helpers.bash, line 608,
#  from function `_test_network_reload' in file ./500-networking.bats, line 332,
#  in test file ./500-networking.bats, line 434)
#   `_test_network_reload pasta' failed
```

...or am I missing something obvious? Should I try to figure that out / help with debugging? In general, let me know if / how I can help here.

**Comment by Honny1** [comment]
> > I am not sure about the packit failure.
> 
> To me the Packit failure (https://dashboard.packit.dev/jobs/testing-farm/1544417 / https://artifacts.dev.testing-farm.io/177feab7-7054-4ff9-9cf3-2f9f3f9ac985/) looks like a failure of a relevant test. From the second link:
> 
> ```
> not ok 619 [500] podman network reload
> # (from function `bail-now' in file ./helpers.bash, line 230,
> #  from function `die' in file ./helpers.bash, line 967,
> #  from function `run_podman' in file ./helpers.bash, line 608,
> #  from function `_test_network_reload' in file ./500-networking.bats, line 332,
> #  in test file ./500-networking.bats, line 434)
> #   `_test_network_reload pasta' failed
> ```
> 
> ...or am I missing something obvious? Should I try to figure that out / help with debugging? In general, let me know if / how I can help here.

It gets a `Permission Denied` error on the Unix socket for pesto. I am not sure if it is a Packit issue or if it is caused by an SELinux policy. I wasn't able to reproduce it locally on my system.

**Comment by sbrivio-rh** [comment]
> It gets a `Permission Denied` error on the Unix socket for pesto.

Oops, so, yes, I missed something obvious. :)

I introduced a new SELinux policy for pesto which explicitly allows `pesto_exec_t` to access `pasta_t` files, but I didn't test that with Podman (just between pesto and stand-alone pasta, pun entirely intended). I'll debug this (perhaps not today though).

**Comment by sbrivio-rh** [comment]
> I introduced a new SELinux policy for pesto which explicitly allows `pesto_exec_t` to access `pasta_t` files, but I didn't test that with Podman (just between pesto and stand-alone pasta, pun entirely intended). I'll debug this (perhaps not today though).

Sorry for the delay. I actually couldn't spot anything wrong in the new profiles, and I'm fairly sure this is a SELinux issue (maybe a relabeling missing as the packages are installed on the CI system), but I looked at all the possible logs and I couldn't find audit logs (/var/log/audit/audit.log) or SELinux warnings anywhere.

Would you have a way to check / extract / find them? I could just spot an out-of-memory message in the kernel log but I don't think it has anything to do with this issue.

**Comment by Luap99** [comment]
> Sorry for the delay. I actually couldn't spot anything wrong in the new profiles, and I'm fairly sure this is a SELinux issue (maybe a relabeling missing as the packages are installed on the CI system), but I looked at all the possible logs and I couldn't find audit logs (/var/log/audit/audit.log) or SELinux warnings anywhere.

we will need to port https://github.com/containers/buildah/pull/6850 here I think for the tmt jobs to capture these logs

**Comment by Honny1** [comment]
> > Sorry for the delay. I actually couldn't spot anything wrong in the new profiles, and I'm fairly sure this is a SELinux issue (maybe a relabeling missing as the packages are installed on the CI system), but I looked at all the possible logs and I couldn't find audit logs (/var/log/audit/audit.log) or SELinux warnings anywhere.
> 
> we will need to port [containers/buildah#6850](https://github.com/containers/buildah/pull/6850) here I think for the tmt jobs to capture these logs

I created PR for that: https://github.com/containers/podman/pull/28737

**Comment by Honny1** [comment]
@Luap99 @sbrivio-rh There seems to be an issue with SELinux. Based on these logs: https://artifacts.dev.testing-farm.io/2e1dcd26-d5e6-484d-854a-233494c648e2/work-local-rootlessa_2kq5lh/plans/system/local-rootless/data/logs/

```
type=AVC msg=audit(1779199621.208:2040): avc:  denied  { create } for  pid=217062 comm="pasta.avx2" name="pasta.sock" scontext=unconfined_u:unconfined_r:pasta_t:s0-s0:c0.c1023 tcontext=unconfined_u:object_r:ifconfig_var_run_t:s0 tclass=sock_file permissive=0
```
cc @lsm5

**Comment by sbrivio-rh** [comment]
> @Luap99 @sbrivio-rh There seems to be an issue with SELinux. Based on these logs: https://artifacts.dev.testing-farm.io/2e1dcd26-d5e6-484d-854a-233494c648e2/work-local-rootlessa_2kq5lh/plans/system/local-rootless/data/logs/
> 
> ```
> type=AVC msg=audit(1779199621.208:2040): avc:  denied  { create } for  pid=217062 comm="pasta.avx2" name="pasta.sock" scontext=unconfined_u:unconfined_r:pasta_t:s0-s0:c0.c1023 tcontext=unconfined_u:object_r:ifconfig_var_run_t:s0 tclass=sock_file permissive=0
> ```

Thanks! Probably my stuff then, I'll keep you posted. Do you need a workaround meanwhile (something subtler than `setenforce 0'`)?

**Comment by Honny1** [comment]
> Thanks! Probably my stuff then, I'll keep you posted. Do you need a workaround meanwhile (something subtler than `setenforce 0'`)?

Thanks, taking care of that. I am not sure if I need a workaround.

**Comment by sbrivio-rh** [comment]
> @Luap99 @sbrivio-rh There seems to be an issue with SELinux. Based on these logs: https://artifacts.dev.testing-farm.io/2e1dcd26-d5e6-484d-854a-233494c648e2/work-local-rootlessa_2kq5lh/plans/system/local-rootless/data/logs/
> 
> ```
> type=AVC msg=audit(1779199621.208:2040): avc:  denied  { create } for  pid=217062 comm="pasta.avx2" name="pasta.sock" scontext=unconfined_u:unconfined_r:pasta_t:s0-s0:c0.c1023 tcontext=unconfined_u:object_r:ifconfig_var_run_t:s0 tclass=sock_file permissive=0
> ```

Fix posted: https://archives.passt.top/passt-dev/20260519173634.3551297-1-sbrivio@redhat.com/

I plan to make a new release rather soon including that. There are already a couple of regressions we just fixed (not touching Podman) or we might be able to fix soon, so a new release in a couple of days makes sense anyway.

File: go.mod [hunk 7]
```diff
@@ -188,3 +188,9 @@ require (
 	gopkg.in/tomb.v1 v1.0.0-20141024135613-dd632973f1e7 // indirect
 	tags.cncf.io/container-device-interface/specs-go v1.1.0 // indirect
 )
+
+replace go.podman.io/common => github.com/containers/container-libs/common v0.67.2-0.20260518190429-32704ef88648
+
+replace go.podman.io/image/v5 => github.com/containers/container-libs/image/v5 v5.0.0-20260518190429-32704ef88648
+
+replace go.podman.io/storage => github.com/containers/container-libs/storage v0.0.0-20260518190429-32704ef88648
```
Comment (Luap99) [inline go.mod]: Do not use replace, just `go get go.podman.io/storage@main && go get go.podman.io/image/v5@main &&  go get go.podman.io/common@main && make vendor`

also need to cleanup the commit description of the vendor commit

File: rpm/podman.spec [hunk 8]
```diff
@@ -111,6 +111,8 @@ Requires: containers-common-extra
 %if %{defined sequoia}
 Requires: podman-sequoia
 %endif
+# pesto binary (ships with passt) is needed for rootless_port_forwarder="pasta"
+Recommends: passt >= 0^20260507.g1afd4ed
```
Comment (Luap99) [inline rpm/podman.spec]: I think it would be best to set the dep where we require passt today https://github.com/containers/container-libs/blob/7e9ee2072844dda92ca80f5a4031f8358c5df2b7/common/rpm/containers-common.spec#L71

Comment (Honny1) [reply rpm/podman.spec]: https://github.com/containers/container-libs/pull/860

**Review by Luap99 — APPROVED** [review]
LGTM

**Comment by TomSweeneyRedHat** [comment]
LGTM
It would be nice to get another set of eyeballs on this.

**Comment by ricardobranco777** [comment]
I think the tests should've gone in `test/system/505-networking-pasta.bats`

**Comment by sbrivio-rh** [comment]
> > @Luap99 @sbrivio-rh There seems to be an issue with SELinux. Based on these logs: https://artifacts.dev.testing-farm.io/2e1dcd26-d5e6-484d-854a-233494c648e2/work-local-rootlessa_2kq5lh/plans/system/local-rootless/data/logs/
> > ```
> > type=AVC msg=audit(1779199621.208:2040): avc:  denied  { create } for  pid=217062 comm="pasta.avx2" name="pasta.sock" scontext=unconfined_u:unconfined_r:pasta_t:s0-s0:c0.c1023 tcontext=unconfined_u:object_r:ifconfig_var_run_t:s0 tclass=sock_file permissive=0
> > ```
> 
> Fix posted: https://archives.passt.top/passt-dev/20260519173634.3551297-1-sbrivio@redhat.com/
> 
> I plan to make a new release rather soon including that. There are already a couple of regressions we just fixed (not touching Podman) or we might be able to fix soon, so a new release in a couple of days makes sense anyway.

Fix merged as https://passt.top/passt/commit/?id=76fd54667ee516e7d6a7ff59befb4a00895b9863, released in [2026_05_26.038c51e](https://archives.passt.top/passt-user/20260527093624.6349111e@elisabeth/) and corresponding Fedora Rawhide [update](https://bodhi.fedoraproject.org/updates/FEDORA-2026-9391bd8f36).

**Comment by sbrivio-rh** [comment]
> I think the tests should've gone in `test/system/505-networking-pasta.bats`

@ricardobranco777 I think in a near future we'll need to add more comprehensive tests checking edge cases and specific functionalities of forwarding to custom network using pasta, see also https://github.com/containers/container-libs/pull/755#issuecomment-4395861122 for context, and those would belong in `505-networking-pasta.bats`.

But for the moment the test, rather simple, needs to run against rootlessport as well as pasta, so that's not really a specific pasta test and I feel like `500-networking.bats` is a better match. What do you think?

**Comment by ricardobranco777** [comment]
> > I think the tests should've gone in `test/system/505-networking-pasta.bats`
> 
> @ricardobranco777 I think in a near future we'll need to add more comprehensive tests checking edge cases and specific functionalities of forwarding to custom network using pasta, see also https://github.com/containers/container-libs/pull/755#issuecomment-4395861122 for context, and those would belong in `505-networking-pasta.bats`.
> 
> But for the moment the test, rather simple, needs to run against rootlessport as well as pasta, so that's not really a specific pasta test and I feel like `500-networking.bats` is a better match. What do you think?

Makes sense.  Thanks.
